back/src/common/repository/helper.ts

```diff
@@ -1,45 +1,62 @@
-import { RepositoryTransaction } from "../../forms/repository/types";
 import prisma from "../../prisma";
+import { RepositoryFnBuilder, RepositoryTransaction } from "./types";
 
-export function transactionWrapper<Builder extends (args) => any>(
-  user: Express.User,
-  transaction: RepositoryTransaction | undefined,
-  builder: Builder
+type Callback = Parameters<RepositoryTransaction["addAfterCommitCallback"]>[0];
+
+type RepositoryContext = {
+  user: Express.User;
+  transaction?: RepositoryTransaction;
+};
+
+/**
+ * Provide a transaction if not provided
+ * and add commit callback to prisma transaction
+ * @param builder
+ * @param context
+ */
+export function transactionWrapper<FnResult>(
+  builder: RepositoryFnBuilder<FnResult>,
+  context: RepositoryContext
 ) {
-  return (...args) => {
-    if (transaction) {
-      return builder({ user, prisma: transaction })(...args);
+  return async (...args) => {
+    const callbacks: Callback[] = [];
+
+    const result = context.transaction
+      ? await builder({
+          user: context.user,
+          prisma: {
+            ...context.transaction,
+            addAfterCommitCallback: callback => {
+              callbacks.push(callback);
+            }
+          }
+        })(...args)
+      : await runInTransaction(async transaction =>
+          builder({
+            user: context.user,
+            prisma: {
+              ...transaction,
+              addAfterCommitCallback: callback => {
+                callbacks.push(callback);
+              }
+            }
+          })(...args)
+        );
+
+    for (const callback of callbacks) {
+      try {
+        await callback();
+      } catch (err) {
+        console.error("Transaction callback error", err);
+      }
     }
 
-    return runInTransaction(newTransaction =>
-      builder({ user, prisma: newTransaction })(...args)
-    );
+    return result;
   };
 }
 
-type Callback = Parameters<RepositoryTransaction["addAfterCommitCallback"]>[0];
-
 export async function runInTransaction<F>(
   func: (transaction: RepositoryTransaction) => Promise<F>
```

Comment:
```suggestion
  func: (transaction: PrismaTransaction) => Promise<F>
```

Si je ne dis pas de bêtise

Reply:
La fonction func prend bien une `RepositoryTransaction` (Cf quelques lignes plus bas). C'est justement le but de wrapper de lui fournir.

Reply:
Avec le revert yes indeed ;)
Après peut être que ton refacto avait du sens: donner accès au hook que dans les repo et pas dans les resolvers.